### DISCUSSION TITLE
[@types/lodash] `omitBy(object, predicate)` – `predicate` should not be `ValueKeyIteratee`?

### DISCUSSION THREAD

**Discussion by aleksi-abr** [DefinitelyTyped/DefinitelyTyped#69674] [Issues with a @types package]
`_.omitBy(object, predicate)` and its variations should be typed so that `predicate` only allows a function type.

```typescript
import _ from 'lodash';

const myObject = {
  a: "b",
  c: 1,
  d: true,
  e: false
};

// works as intended
console.log("test1", _.omitBy(myObject, (v, k) => k !== "a"));  // returns { a: "b" }

// works, but should it?
console.log("test2", _.omitBy(myObject, 0));  // returns { c: 1, d: true, e: false }
console.log("test3", _.omitBy(myObject, null)); // returns { e: false }
console.log("test4", _.omitBy(myObject, {})); // returns {}

// works, but meaningless; all return the complete object
console.log("test5", _.omitBy(myObject, "a")); 
console.log("test6", _.omitBy(myObject, "b"));
console.log("test7", _.omitBy(myObject, "hello"));
console.log("test8", _.omitBy(myObject, 1234));
console.log("test9", _.omitBy(myObject, false));
console.log("test10", _.omitBy(myObject, { a: true, c: false, d: null }));
console.log("test11", _.omitBy(myObject, []));
console.log("test12", _.omitBy(myObject, [0, 1, 2]));
console.log("test13", _.omitBy(myObject, ["a", "b"]));
```

The lax typing allows an easy error that TypeScript could easily catch – passing a primitive instead of a function as `predicate`. The root cause is that [`ValueKeyIteratee`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/lodash/common/common.d.ts#L240) allows both a function type **or** an `IterateeShorthand`. I'm guessing in the case of `_.omitBy()` only the function type should be allowed?

At minimum, since `string` and `boolean` do not appear to do anything, those types should not be allowed in `predicate`?

Although some other values work for `predicate`, [per documentation](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/lodash/common/object.d.ts#L2008) it is expected to be a function.

[a comment by an automated account is not shown]

**Comment by aj-r** [comment]
Lodash iteratee arguments accept any argument that _.iteratee accepts:
https://lodash.com/docs/4.17.15#iteratee. Hence it needs to accept strings,
objects, and arrays. Perhaps numbers shouldn't be valid though.
